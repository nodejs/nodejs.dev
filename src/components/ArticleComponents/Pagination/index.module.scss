.pagination {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  list-style: none;
  margin: 0;
  padding: 5rem 0 0 0;

  a {
    color: var(--black7) !important;
    font-family: var(--sans-serif);
    font-size: 1.4rem;
    font-weight: var(--font-weight-regular);
    text-decoration: none !important;
    text-transform: uppercase;
    vertical-align: middle;

    &:hover {
      color: var(--brand-light) !important;
    }
  }
}
