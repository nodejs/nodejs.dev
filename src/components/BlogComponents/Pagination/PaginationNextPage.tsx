import React from 'react';
import classnames from 'classnames';
import { LocalizedLink as Link } from 'gatsby-theme-i18n';
import { FormattedMessage } from 'react-intl';
import styles from './index.module.scss';

interface Props {
  className: string;
  currentPage: number;
  hrefBuilder: (n: number) => string;
  onPageChange: (n: number) => void;
  pageCount: number;
}

const PaginationNextPage = ({
  className,
  currentPage,
  hrefBuilder,
  onPageChange,
  pageCount,
}: Props) => {
  const componentClassName = classnames(className, styles.next);
  const content = <FormattedMessage id="components.pagination.next" />;
  const disabled = currentPage === pageCount;

  if (disabled) {
    return (
      <span aria-disabled="true" className={componentClassName}>
        {content}
      </span>
    );
  }

  const nextPage = currentPage + 1;

  return (
    <Link
      aria-label="Next Page"
      onClick={onPageChange(nextPage)}
      className={componentClassName}
      rel="next"
      to={hrefBuilder(nextPage)}
    >
      {content}
    </Link>
  );
};

export default PaginationNextPage;
