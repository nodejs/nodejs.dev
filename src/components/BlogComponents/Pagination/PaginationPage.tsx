import React, { MouseEvent } from 'react';
import classnames from 'classnames';
import { LocalizedLink as Link } from 'gatsby-theme-i18n';
import { PageModel } from './util/generatePageList';

interface Props {
  className: string;
  page: PageModel;
  hrefBuilder: (n: number) => string;
  onPageChange?: (e: MouseEvent) => void;
}

const PaginationPage = ({
  className,
  page,
  hrefBuilder,
  onPageChange,
}: Props) => {
  const content = String(page.num);
  const componentClassName = classnames(className, page?.className ?? '');

  if (page.selected) {
    return (
      <em aria-current="page" className={componentClassName}>
        {content}
      </em>
    );
  }

  return (
    <Link
      aria-label={`Page ${page.num}`}
      onClick={onPageChange}
      className={componentClassName}
      to={hrefBuilder(page.num)}
    >
      {content}
    </Link>
  );
};

export default PaginationPage;
