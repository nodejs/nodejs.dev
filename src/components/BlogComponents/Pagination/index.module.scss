.pagination {
  display: flex;
  flex-wrap: wrap;
  list-style: none;
  margin: 0;
  padding: 0;

  .pagination-page {
    align-items: center;
    background-color: var(--brand0);
    color: var(--black9);
    display: inline-flex;
    font-size: var(--font-size-caption);
    font-style: normal;
    justify-content: center;
    line-height: var(--line-height-caption);
    padding: var(--space-07) var(--space-13);
    text-align: center;
    text-decoration: none;
    transition: background-color 0.2s ease-in-out;
    user-select: none;
    white-space: nowrap;

    &:hover,
    &:focus {
      background-color: var(--brand10);
      outline: 0;
      text-decoration: none;
      transition-duration: 0.1s;
    }

    &.prev,
    &.next {
      background-color: var(--black2);
      color: var(--black9);
    }

    &[aria-current],
    &[aria-current]:hover {
      background-color: var(--color-brand-primary);
      color: var(--black0);
    }

    &.prev {
      border-radius: 4px 0 0 4px;
    }

    &.next {
      border-radius: 0 4px 4px 0;
    }

    &[aria-disabled],
    &[aria-disabled]:hover {
      background-color: var(--black1);
      color: var(--black9);
      cursor: not-allowed;
    }
  }
}

:global(.dark) {
  .pagination-page {
    color: var(--black0);

    &:hover,
    &:focus {
      color: var(--black9);
    }
  }
}
