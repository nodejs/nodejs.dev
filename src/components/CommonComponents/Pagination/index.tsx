import React, { FC, MouseEvent, useCallback, useMemo } from 'react';
import { buildPaginationModel } from './util/buildPaginationModel';
import Page from './Page';
import styles from './index.module.scss';

interface Props {
  currentPage: number;
  hrefBuilder?: (n: number) => string;
  marginPageCount: number;
  pageCount: number;
  onPageChange?: (e: MouseEvent, n: number) => void;
  showPages?: boolean;
  surroundingPageCount?: number;
  wrapperClassName?: string;
}

const defaultHrefBuilder = (pageNum: number): string => `#${pageNum}`;

const Pagination: FC<Props> = ({
  pageCount,
  currentPage,
  onPageChange,
  hrefBuilder = defaultHrefBuilder,
  marginPageCount = 1,
  showPages = true,
  surroundingPageCount = 2,
  wrapperClassName,
}) => {
  const pageChange = useCallback(
    (n: number) => (e: MouseEvent) => onPageChange && onPageChange(e, n),
    [onPageChange]
  );

  const model = useMemo(
    () =>
      buildPaginationModel(
        pageCount,
        currentPage,
        !!showPages,
        marginPageCount,
        surroundingPageCount
      ),
    [pageCount, currentPage, showPages, marginPageCount, surroundingPageCount]
  );

  const pageElements = useMemo(
    () =>
      model.map(page => (
        <li key={`${page.type}${page.num}`}>
          <Page
            className={styles.paginationPage}
            hrefBuilder={hrefBuilder}
            page={page}
            onPageChange={pageChange(page.num)}
          />
        </li>
      )),
    [model, hrefBuilder, pageChange]
  );

  return (
    <nav aria-label="Pagination" className={wrapperClassName}>
      <ul className={styles.pagination}>{pageElements}</ul>
    </nav>
  );
};

export default Pagination;
