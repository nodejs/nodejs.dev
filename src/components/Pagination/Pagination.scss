.link {
  color: var(--black7) !important;
  text-transform: uppercase;
  text-decoration: none !important;
  font-size: 1.4rem;
  font-weight: normal;
  font-family: var(--sans-serif);
  vertical-align: middle;
  &:hover {
    color: var(--brand-light) !important;
  }
}

.pagination {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  list-style: none;
  padding: 5rem 0;
}
