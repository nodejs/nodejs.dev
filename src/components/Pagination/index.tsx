import { Link } from 'gatsby';
import React from 'react';
import { PaginationInfo } from '../../types';
import './Pagination.scss';

interface Props {
  previous?: PaginationInfo;
  next?: PaginationInfo;
}

const Pagination = ({ previous, next }: Props): JSX.Element | null => {
  if (!previous && !next) return null;

  return (
    <ul className="pagination">
      <li>
        {previous && previous.title && (
          <Link className="link" to={`/learn/${previous.slug}`} rel="prev">
            ← &nbsp; Prev
          </Link>
        )}
      </li>
      <li>
        {next && next.title && (
          <Link className="link" to={`/learn/${next.slug}`} rel="next">
            Next &nbsp; →
          </Link>
        )}
      </li>
    </ul>
  );
};

export default Pagination;
