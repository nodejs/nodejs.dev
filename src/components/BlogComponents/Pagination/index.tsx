import React, { MouseEvent, useCallback } from 'react';
import { generatePageList } from './util/generatePageList';
import { PageType } from '../../../types/pagination';
import PaginationPreviousPage from './PaginationPreviousPage';
import PaginationNextPage from './PaginationNextPage';
import PaginationPage from './PaginationPage';
import styles from './index.module.scss';

interface Props {
  currentPage: number;
  hrefBuilder?: (n: number) => string;
  pageCount: number;
  onPageChange?: (e: MouseEvent, n: number) => void;
  showPages?: boolean;
  wrapperClassName?: string;
}

const defaultHrefBuilder = (pageNum: number): string => `#${pageNum}`;

const Pagination = ({
  pageCount,
  currentPage,
  onPageChange,
  hrefBuilder = defaultHrefBuilder,
  showPages = true,
  wrapperClassName,
}: Props) => {
  const pageChange = useCallback(
    (n: number) => (e: MouseEvent) => onPageChange && onPageChange(e, n),
    [onPageChange]
  );

  const pageList = generatePageList(pageCount, currentPage, !!showPages);

  const pageElements = pageList.map(page => (
    <li key={`${page.type}${page.num}`}>
      <PaginationPage
        className={styles.paginationPage}
        hrefBuilder={hrefBuilder}
        page={page}
        onPageChange={pageChange(page.num)}
      />
    </li>
  ));

  return (
    <nav aria-label="Pagination" className={wrapperClassName}>
      <ul className={styles.pagination}>
        <li key={PageType.PREV}>
          <PaginationPreviousPage
            className={styles.paginationPage}
            currentPage={currentPage}
            hrefBuilder={hrefBuilder}
            onPageChange={pageChange}
          />
        </li>
        {pageElements}
        <li key={PageType.NEXT}>
          <PaginationNextPage
            className={styles.paginationPage}
            currentPage={currentPage}
            hrefBuilder={hrefBuilder}
            onPageChange={pageChange}
            pageCount={pageCount}
          />
        </li>
      </ul>
    </nav>
  );
};

export default Pagination;
