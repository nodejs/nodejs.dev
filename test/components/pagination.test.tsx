import React from 'react';
import renderer from 'react-test-renderer';
import Pagination from '../../src/components/pagination';

describe('Pagination component', () => {
  it('renders links to the next and previous page', () => {
    const next = { frontmatter: { title: 'next-page' }, fields: { slug: 'next-page' }};
    const prev = { frontmatter: { title: 'prev-page' }, fields: { slug: 'prev-page' }};
    const tree = renderer.create(<Pagination next={next} previous={prev} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders correctly when there is no next page', () => {
    const prev = { frontmatter: { title: 'prev-page' }, fields: { slug: 'prev-page' }};
    const tree = renderer.create(<Pagination previous={prev} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders correctly when there is no previous page', () => {
    const next = { frontmatter: { title: 'next-page' }, fields: { slug: 'next-page' }};
    const tree = renderer.create(<Pagination next={next} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('only renders links to pages that has a title', () => {
    const next = { frontmatter: { title: 'next-page' }, fields: { slug: 'next-page' }};
    const prev = { frontmatter: { title: null }, fields: { slug: 'prev-page' }};
    const tree = renderer.create(<Pagination next={next} previous={prev} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
