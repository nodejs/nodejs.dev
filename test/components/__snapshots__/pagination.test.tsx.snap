// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component only renders links to pages that has a title 1`] = `
<ul
  style={
    Object {
      "display": "flex",
      "flexWrap": "wrap",
      "justifyContent": "space-between",
      "listStyle": "none",
      "padding": "30px",
    }
  }
>
  <li />
  <li>
    <a
      href="/learn/next-page"
      rel="next"
    >
      Next   →
    </a>
  </li>
</ul>
`;

exports[`Pagination component renders correctly when there is no next page 1`] = `
<ul
  style={
    Object {
      "display": "flex",
      "flexWrap": "wrap",
      "justifyContent": "space-between",
      "listStyle": "none",
      "padding": "30px",
    }
  }
>
  <li>
    <a
      href="/learn/prev-page"
      rel="prev"
    >
      ←   Prev
    </a>
  </li>
  <li />
</ul>
`;

exports[`Pagination component renders correctly when there is no previous page 1`] = `
<ul
  style={
    Object {
      "display": "flex",
      "flexWrap": "wrap",
      "justifyContent": "space-between",
      "listStyle": "none",
      "padding": "30px",
    }
  }
>
  <li />
  <li>
    <a
      href="/learn/next-page"
      rel="next"
    >
      Next   →
    </a>
  </li>
</ul>
`;

exports[`Pagination component renders links to the next and previous page 1`] = `
<ul
  style={
    Object {
      "display": "flex",
      "flexWrap": "wrap",
      "justifyContent": "space-between",
      "listStyle": "none",
      "padding": "30px",
    }
  }
>
  <li>
    <a
      href="/learn/prev-page"
      rel="prev"
    >
      ←   Prev
    </a>
  </li>
  <li>
    <a
      href="/learn/next-page"
      rel="next"
    >
      Next   →
    </a>
  </li>
</ul>
`;
